File: flut/flutter/widgets/shortcuts.py
from typing import Callable, Optional, override

from flut._flut.engine import FlutEnumObject, FlutValueObject, _flut_pack_value
from flut._flut.runtime import _flut_unpack_required_field
from flut.flutter.services.keyboard_key import LogicalKeyboardKey
from flut.flutter.widgets.framework import Widget


class LockState(FlutEnumObject):
    ignored: "LockState"
    locked: "LockState"
    unlocked: "LockState"


class ShortcutActivator:
    def __init__(self):
        pass


class SingleActivator(FlutValueObject):
    _flut_type = "SingleActivator"

    def __init__(
        self,
        trigger: LogicalKeyboardKey,
        *,
        control: bool = False,
        shift: bool = False,
        alt: bool = False,
        meta: bool = False,
        numLock: LockState = LockState.ignored,
        includeRepeats: bool = True,
    ):
        super().__init__()
        self.trigger = trigger
        self.control = control
        self.shift = shift
        self.alt = alt
        self.meta = meta
        self.numLock = numLock
        self.includeRepeats = includeRepeats

    @staticmethod
    def _flut_unpack(data: dict) -> "SingleActivator":
        return SingleActivator(
            trigger=_flut_unpack_required_field(data, "trigger"),
            control=_flut_unpack_required_field(data, "control"),
            shift=_flut_unpack_required_field(data, "shift"),
            alt=_flut_unpack_required_field(data, "alt"),
            meta=_flut_unpack_required_field(data, "meta"),
            numLock=_flut_unpack_required_field(data, "numLock"),
            includeRepeats=_flut_unpack_required_field(data, "includeRepeats"),
        )

    @override
    def _flut_pack(self) -> dict:
        result = self._flut_base_props()
        result["trigger"] = _flut_pack_value(self.trigger)
        result["control"] = _flut_pack_value(self.control)
        result["shift"] = _flut_pack_value(self.shift)
        result["alt"] = _flut_pack_value(self.alt)
        result["meta"] = _flut_pack_value(self.meta)
        result["numLock"] = _flut_pack_value(self.numLock)
        result["includeRepeats"] = _flut_pack_value(self.includeRepeats)
        return result


class CharacterActivator(FlutValueObject):
    _flut_type = "CharacterActivator"

    def __init__(
        self,
        character: str,
        *,
        control: bool = False,
        meta: bool = False,
        alt: bool = False,
        includeRepeats: bool = True,
    ):
        super().__init__()
        self.character = character
        self.control = control
        self.meta = meta
        self.alt = alt
        self.includeRepeats = includeRepeats

    @staticmethod
    def _flut_unpack(data: dict) -> "CharacterActivator":
        return CharacterActivator(
            character=_flut_unpack_required_field(data, "character"),
            control=_flut_unpack_required_field(data, "control"),
            meta=_flut_unpack_required_field(data, "meta"),
            alt=_flut_unpack_required_field(data, "alt"),
            includeRepeats=_flut_unpack_required_field(data, "includeRepeats"),
        )

    @override
    def _flut_pack(self) -> dict:
        result = self._flut_base_props()
        result["character"] = _flut_pack_value(self.character)
        result["control"] = _flut_pack_value(self.control)
        result["meta"] = _flut_pack_value(self.meta)
        result["alt"] = _flut_pack_value(self.alt)
        result["includeRepeats"] = _flut_pack_value(self.includeRepeats)
        return result


class Shortcuts(Widget):
    _flut_type = "Shortcuts"

    def __init__(
        self,
        *,
        key=None,
        shortcuts: dict,
        debugLabel: Optional[str] = None,
        includeSemantics: bool = True,
        child: Widget,
    ):
        super().__init__(key=key)
        self.shortcuts = shortcuts
        self.debugLabel = debugLabel
        self.includeSemantics = includeSemantics
        self.child = child

    @override
    def _flut_pack(self) -> dict:
        result = self._flut_base_props()
        shortcuts_list = []
        for activator, intent in self.shortcuts.items():
            entry = {
                "activator": _flut_pack_value(activator),
                "intentName": type(intent).__name__,
            }
            data = {}
            for k, v in intent.__dict__.items():
                if not k.startswith("_"):
                    data[k] = _flut_pack_value(v)
            if data:
                entry["intentData"] = data
            shortcuts_list.append(entry)
        result["shortcuts"] = shortcuts_list
        if self.debugLabel is not None:
            result["debugLabel"] = _flut_pack_value(self.debugLabel)
        result["includeSemantics"] = _flut_pack_value(self.includeSemantics)
        result["child"] = _flut_pack_value(self.child)
        return result


class CallbackShortcuts(Widget):
    _flut_type = "CallbackShortcuts"

    def __init__(self, *, key=None, bindings: dict, child: Widget):
        super().__init__(key=key)
        self.bindings = bindings
        self.child = child

    @override
    def _flut_pack(self) -> dict:
        result = self._flut_base_props()
        bindings_list = []
        for activator, callback in self.bindings.items():
            cb = self._register_action(callback, "VoidCallback")
            bindings_list.append(
                {
                    "activator": _flut_pack_value(activator),
                    "callback": cb._flut_pack(),
                }
            )
        result["bindings"] = bindings_list
        result["child"] = _flut_pack_value(self.child)
        return result

File: flut/.flutter/lib/flutter/widgets/shortcuts.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flut/flut/runtime.dart';
import 'package:flut/flut/object.dart';
import 'package:flut/flutter/widgets/actions.dart';

class FlutLockState extends FlutEnumObject<LockState> {
  const FlutLockState()
    : super('LockState', const {
        'ignored': LockState.ignored,
        'locked': LockState.locked,
        'unlocked': LockState.unlocked,
      });
}

class FlutSingleActivator extends FlutValueObject {
  final SingleActivator activator;
  const FlutSingleActivator(this.activator) : super('SingleActivator');

  @override
  Map<String, dynamic> flutEncode() {
    final result = flutBaseProps();
    result['trigger'] = encodeLogicalKey(activator.trigger);
    result['control'] = activator.control;
    result['shift'] = activator.shift;
    result['alt'] = activator.alt;
    result['meta'] = activator.meta;
    result['numLock'] = const FlutLockState().flutEncode(activator.numLock);
    result['includeRepeats'] = activator.includeRepeats;
    return result;
  }

  static Map<String, dynamic> encodeLogicalKey(LogicalKeyboardKey key) {
    return {'_flut_type': 'LogicalKeyboardKey', 'keyId': key.keyId};
  }

  static SingleActivator? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return SingleActivator(
      runtime.unpackRequiredField<LogicalKeyboardKey>(data, 'trigger'),
      control: runtime.unpackRequiredField<bool>(data, 'control'),
      shift: runtime.unpackRequiredField<bool>(data, 'shift'),
      alt: runtime.unpackRequiredField<bool>(data, 'alt'),
      meta: runtime.unpackRequiredField<bool>(data, 'meta'),
      numLock: runtime.unpackRequiredField<LockState>(data, 'numLock'),
      includeRepeats: runtime.unpackRequiredField<bool>(data, 'includeRepeats'),
    );
  }
}

class FlutCharacterActivator extends FlutValueObject {
  final CharacterActivator activator;
  const FlutCharacterActivator(this.activator) : super('CharacterActivator');

  @override
  Map<String, dynamic> flutEncode() {
    final result = flutBaseProps();
    result['character'] = activator.character;
    result['control'] = activator.control;
    result['meta'] = activator.meta;
    result['alt'] = activator.alt;
    result['includeRepeats'] = activator.includeRepeats;
    return result;
  }

  static CharacterActivator? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return CharacterActivator(
      runtime.unpackRequiredField<String>(data, 'character'),
      control: runtime.unpackRequiredField<bool>(data, 'control'),
      meta: runtime.unpackRequiredField<bool>(data, 'meta'),
      alt: runtime.unpackRequiredField<bool>(data, 'alt'),
      includeRepeats: runtime.unpackRequiredField<bool>(data, 'includeRepeats'),
    );
  }
}

/// Bridge intent used by FlutShortcuts to link Shortcuts → FlutActionScope.
class _FlutShortcutIntent extends Intent {
  final String intentName;
  final Map<String, dynamic>? intentData;
  const _FlutShortcutIntent(this.intentName, [this.intentData]);
}

class FlutShortcuts {
  FlutShortcuts._();

  static Widget? flutDecode(FlutRuntime runtime, Map<String, dynamic> data) {
    final shortcutsList = data['shortcuts'] as List<dynamic>;
    final Map<ShortcutActivator, Intent> shortcuts = {};
    for (final item in shortcutsList) {
      final itemMap = item as Map<String, dynamic>;
      final activatorData = itemMap['activator'] as Map<String, dynamic>;
      final activator = _decodeActivator(runtime, activatorData);
      final intentName = itemMap['intentName'] as String;
      final intentData = itemMap['intentData'] as Map<String, dynamic>?;
      if (activator != null) {
        shortcuts[activator] = _FlutShortcutIntent(intentName, intentData);
      }
    }

    return Shortcuts(
      key: runtime.decodeKey(data),
      shortcuts: shortcuts,
      debugLabel: runtime.unpackOptionalField<String>(data, 'debugLabel'),
      includeSemantics: runtime.unpackRequiredField<bool>(
        data,
        'includeSemantics',
      ),
      child: Actions(
        actions: <Type, Action<Intent>>{
          _FlutShortcutIntent: CallbackAction<_FlutShortcutIntent>(
            onInvoke: (intent) {
              final ctx = primaryFocus?.context;
              if (ctx != null) {
                FlutActionScope.invokeHandler(
                  ctx,
                  runtime,
                  intent.intentName,
                  intent.intentData,
                );
              }
              return null;
            },
          ),
        },
        child: runtime.unpackRequiredField<Widget>(data, 'child'),
      ),
    );
  }
}

class FlutCallbackShortcuts {
  FlutCallbackShortcuts._();

  static Widget? flutDecode(FlutRuntime runtime, Map<String, dynamic> data) {
    final bindingsList = data['bindings'] as List<dynamic>;
    final Map<ShortcutActivator, VoidCallback> bindings = {};
    for (final item in bindingsList) {
      final itemMap = item as Map<String, dynamic>;
      final activatorData = itemMap['activator'] as Map<String, dynamic>;
      final activator = _decodeActivator(runtime, activatorData);
      final callbackData = itemMap['callback'] as Map<String, dynamic>;
      final callableRef = FlutCallableRef.flutDecode(runtime, callbackData);
      if (activator != null && callableRef != null) {
        final adapted = runtime.adaptCallableByType(callableRef);
        bindings[activator] = adapted as VoidCallback;
      }
    }

    return CallbackShortcuts(
      key: runtime.decodeKey(data),
      bindings: bindings,
      child: runtime.unpackRequiredField<Widget>(data, 'child'),
    );
  }
}

/// Decode a ShortcutActivator from wire data.
ShortcutActivator? _decodeActivator(
  FlutRuntime runtime,
  Map<String, dynamic> data,
) {
  final type = data['_flut_type'] as String?;
  switch (type) {
    case 'SingleActivator':
      return FlutSingleActivator.flutDecode(runtime, data);
    case 'CharacterActivator':
      return FlutCharacterActivator.flutDecode(runtime, data);
    default:
      return null;
  }
}
